File: merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> (a[0]-b[0]));
        int[] curr = intervals[0];
        List<int[]> result = new ArrayList<>();
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] > curr[1]) {
                result.add(new int[] {curr[0], curr[1]});
                curr[0] = intervals[i][0];
                curr[1] = intervals[i][1];
            } else {
                curr[1] = Math.max(curr[1], intervals[i][1]);
            }
        }
        result.add(new int[] {curr[0], curr[1]});
        return result.toArray(int[][]::new);
    }
}

File: merge-intervals/merge-intervals.go
func merge(intervals [][]int) (result [][]int) {
    sort.Sort(interval(intervals))
    
    start := intervals[0][0]
    end := intervals[0][1]
    
    for i := 1; i < len(intervals); i++ {
        if intervals[i][0] <= end {
            end = max(end, intervals[i][1])
        } else {
            result = append(result, []int{start, end})
            start = intervals[i][0]
            end = intervals[i][1]
        }
    }
    
    result = append(result, []int{start, end})
    
    return
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

type interval [][]int

func (i interval) Len() int {
    return len(i)
}

func (i interval) Swap(a, b int) {
    i[a], i[b] = i[b], i[a]
}

func (i interval) Less(a, b int) bool {
    return i[a][0] < i[b][0]
}